CM7/Core/Src/gatedriver.c

```diff
@@ -81,10 +90,62 @@ int16_t gatedrv_read_dc_current(gatedriver_t* drv)
 }
 
 /* Note: This has to atomically write to ALL PWM registers */
-//TODO: mechanism for PWM synchronization
-int16_t gatedrv_write_pwm(gatedriver_t* drv)
+int16_t gatedrv_write_pwm(gatedriver_t* drv, float dutyCycles[])
 {
-
+	// computing pulses
+	uint32_t pulses[3];
+	pulses[0] = (uint32_t) (dutyCycles[0] * PERIOD_VALUE / 100);
+	pulses[1] = (uint32_t) (dutyCycles[1] * PERIOD_VALUE / 100);
+	pulses[2] = (uint32_t) (dutyCycles[2] * PERIOD_VALUE / 100);
+
+	// Common configuration for all channels
+	TIM_OC_InitTypeDef PWMConfig;
+	PWMConfig.OCMode       = TIM_OCMODE_PWM1;
+	PWMConfig.OCPolarity   = TIM_OCPOLARITY_HIGH;
+	PWMConfig.OCNPolarity  = TIM_OCNPOLARITY_HIGH;
+	PWMConfig.OCIdleState  = TIM_OCIDLESTATE_SET;
+	PWMConfig.OCNIdleState = TIM_OCNIDLESTATE_RESET;
+	PWMConfig.OCFastMode   = TIM_OCFAST_DISABLE;
+
+	// Attempting to set channel 1
+	PWMConfig.Pulse = pulses[0];
+	if(HAL_TIM_PWM_ConfigChannel(drv->tim, &PWMConfig, TIM_CHANNEL_1) != HAL_OK)
+	{
+		// do nothing and return
+		return 1;
+	}
+
+	// Attempting to set channel 2
+	PWMConfig.Pulse = pulses[1];
+	if(HAL_TIM_PWM_ConfigChannel(drv->tim, &PWMConfig, TIM_CHANNEL_2) != HAL_OK)
+	{
+		// attempt to revert last channel change and return
+		PWMConfig.Pulse = drv->pulses[0];
+		HAL_TIM_PWM_ConfigChannel(drv->tim, &PWMConfig, TIM_CHANNEL_1);
+
+		return 1;
+	}
+
+	// Attempting to set channel 3
+	PWMConfig.Pulse = pulses[2];
+	if(HAL_TIM_PWM_ConfigChannel(drv->tim, &PWMConfig, TIM_CHANNEL_3) != HAL_OK)
+	{
+		// attempt to revert previous channel changes and return
+		PWMConfig.Pulse = drv->pulses[0];
+		HAL_TIM_PWM_ConfigChannel(drv->tim, &PWMConfig, TIM_CHANNEL_1);
+
+		PWMConfig.Pulse = drv->pulses[1];
+		HAL_TIM_PWM_ConfigChannel(drv->tim, &PWMConfig, TIM_CHANNEL_2);
+
+		return 1;
+	}
+
+	// successful PWM modifications - save configuration and return
+	drv->pulses[0] = pulses[0];
```

Comment:
This bit was to save the old duty cycles in case we errored out in one of the PWM pulse set operations - essentially, if the second set operation failed after the first succeeded, it would backtrack and try to set the first pwm duty cycle back to what it used to be. Similar thing with the third PWM set. If that's unnecessary, I can change it to just halt and return an error code and get rid of this field.

Reply:
Nah I support